stats: create non-trivial ParentHistogram for IsolatedStoreImpl

mergeconflict opened the pull request:
Description: Create a non-trivial ParentHistogramImpl for IsolatedStoreImpl. This replaces HistogramImpl and allows tests to verify that histograms are being written as expected, in the same style as is currently done for counters and gauges.
Risk Level: low
Testing: added unit test; all existing tests pass.
Docs Changes: n/a
Release Notes: n/a
Fixes #6582

Signed-off-by: Dan Rosen <mergeconflict@google.com>

mergeconflict commented:
/review @jmarantz @mattklein123

mergeconflict commented on source/common/stats/thread_local_store.h:
```diff
@@ -69,11 +69,11 @@ class TlsScope;
  * Log Linear Histogram implementation that is stored in the main thread.
  */
-class ParentHistogramImpl : public ParentHistogram, public MetricImpl {
+class ThreadLocalParentHistogramImpl : public ParentHistogram, public MetricImpl {
```
Renamed. This is more consistent with the other `ThreadLocal` classes here, and it lets me grab the `ParentHistogramImpl` name for my not-thread-local variant.

jmarantz commented on source/common/stats/histogram_impl.cc:
```diff
@@ -79,5 +79,64 @@ void HistogramStatisticsImpl::refresh(const histogram_t* new_histogram_ptr) {
+ParentHistogramImpl::ParentHistogramImpl(const std::string& name, Store& parent,
+                                         std::string&& tag_extracted_name, std::vector<Tag>&& tags)
+    : MetricImpl(std::move(tag_extracted_name), std::move(tags)), parent_(parent),
+      interval_histogram_(hist_alloc()), cumulative_histogram_(hist_alloc()),
```
You can also use std::unique_ptr<> with a custom deleter, so you can get rid of the dtor.

mergeconflict replied:
Yeah, there's a `CSmartPtr` helper for this very thing. I didn't use it here just to stay consistent with `ThreadLocalParentHistogram`, which also doesn't use it, but perhaps they both should.

mergeconflict replied:
Actually, decided not to change this. Too many `.get()`s.

jmarantz commented on source/common/stats/histogram_impl.cc:
```diff
@@ -79,5 +79,64 @@ void HistogramStatisticsImpl::refresh(const histogram_t* new_histogram_ptr) {
+    }
+    return absl::StrJoin(summary, " ");
+  } else {
+    return std::string("No recorded values");
```
nit: you can return "No recorded values" here rather than explicitly constructing the string.

mergeconflict replied:
This is what I get for copy-pasting code :)

jmarantz commented on source/common/stats/histogram_impl.cc:
```diff
@@ -79,5 +79,64 @@ void HistogramStatisticsImpl::refresh(const histogram_t* new_histogram_ptr) {
+
+// TODO(mergeconflict): figure out the best way to eliminate duplication between this and
+// bucketSummary below, and the implementations in ThreadLocalParentHistogramImpl.
+const std::string ParentHistogramImpl::quantileSummary() const {
```
nit: I don't usually qualify a by-value return object with 'const' -- I think it doesn't really add anything.

mergeconflict replied:
Yeah, but it's defined that way in the `ParentHistogram` interface.

jmarantz commented on source/common/stats/histogram_impl.cc:
```diff
@@ -79,5 +79,64 @@ void HistogramStatisticsImpl::refresh(const histogram_t* new_histogram_ptr) {
+  }
+}
+
+const std::string ParentHistogramImpl::bucketSummary() const {
```
Per above I don't think you need the 'const' in 'const std::string".

jmarantz commented on test/common/stats/histogram_impl_test.cc:
```diff
@@ -0,0 +1,61 @@
+
+} // namespace
+} // namespace Stats
+} // namespace Envoy
```
nit: add newline

jmarantz commented on test/common/stats/histogram_impl_test.cc:
```diff
@@ -0,0 +1,61 @@
+namespace {
+
+TEST(HistogramImplTest, Basic) {
+  MockStore store;
```
Can you make a test-class that captures these 4 lines of declaration? Then (most) every stanza can be a test method.

mergeconflict replied:
Each "stanza" builds on the previous, they don't really stand alone. The only one that could be its own test is the initial portion about names. But I could buy the "one assert per test" argument too... Do you have strong feelings about this?

mergeconflict commented on test/common/stats/histogram_impl_test.cc:
```diff
@@ -0,0 +1,61 @@
+                            &cumulativeStats = histogram.cumulativeStatistics();
+
+  // A histogram should retain the name it was given.
+  ASSERT_EQ("test", histogram.name());
```
Oops, good catch.

mergeconflict commented:
/wait

mergeconflict commented on include/envoy/stats/histogram.h:
```diff
@@ -17,16 +17,6 @@ class HistogramStatistics {
 public:
   virtual ~HistogramStatistics() {}
 
-  /**
```
These were only used in tests, to equality-compare `HistogramStatistics` objects.

mergeconflict commented on include/envoy/stats/histogram.h:
```diff
@@ -105,16 +95,6 @@ class ParentHistogram : public virtual Histogram {
    * Returns the cumulative histogram summary statistics.
    */
   virtual const HistogramStatistics& cumulativeStatistics() const PURE;
-
```
`quantileSummary` was only used in admin.cc, and `bucketSummary` was unused entirely.

mergeconflict commented on source/common/stats/thread_local_store.cc:
```diff
@@ -558,44 +559,12 @@ void ParentHistogramImpl::merge() {
   }
 }
 
-const std::string ParentHistogramImpl::quantileSummary() const {
```
moved to admin.cc

mergeconflict commented on source/common/stats/thread_local_store.cc:
```diff
@@ -558,44 +559,12 @@ void ParentHistogramImpl::merge() {
-  }
-}
-
-const std::string ParentHistogramImpl::bucketSummary() const {
```
unused

mattklein123 commented:
@mergeconflict QQ: Did you consider adding this functionality only to a derived `TestIsolatedStatsStore` class? Since we don't use this currently in production code I'm mainly wondering what overhead there might be since we will instantiate one of these for every host that Envoy knows about.

mergeconflict commented:
> @mergeconflict QQ: Did you consider adding this functionality only to a derived `TestIsolatedStatsStore` class? Since we don't use this currently in production code I'm mainly wondering what overhead there might be since we will instantiate one of these for every host that Envoy knows about.

Hmm. Yeah, I didn't consider that. It could be significant overhead.

There actually exists a `TestIsolatedStoreImpl` class already, that just contains an `IsolatedStoreImpl`, and wraps each method in a `LockGuard`. I could try to plug in there, but it's a little gnarly. I'll think about it some more.

/wait

jmarantz commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -72,6 +72,13 @@ class HistogramWrapper {
 
+namespace {
+void expectStatsEq(const Stats::HistogramStatistics& lhs, const Stats::HistogramStatistics& rhs) {
+  EXPECT_EQ(lhs.supportedBuckets(), rhs.supportedBuckets());
```
Tiny nit, but It's usually a little better to put the EXPECT calls directly in the test body so that an error points you unambiguously to the test & line.

however I also  love factoring stuff out, which your design here does nicely.

As a compromise, you could pass in a string annotation to this function and use different args at the call-sites to disambiguate should there be a failure. Then you just write:

 EXPECT_EQ(....) << annotation;

mergeconflict replied:
I should probably just define `operator ==` for `HistogramStatistics`.

mergeconflict commented:
> > @mergeconflict QQ: Did you consider adding this functionality only to a derived `TestIsolatedStatsStore` class? Since we don't use this currently in production code I'm mainly wondering what overhead there might be since we will instantiate one of these for every host that Envoy knows about.
> 
> Hmm. Yeah, I didn't consider that. It could be significant overhead.
> 
> There actually exists a `TestIsolatedStoreImpl` class already, that just contains an `IsolatedStoreImpl`, and wraps each method in a `LockGuard`. I could try to plug in there, but it's a little gnarly. I'll think about it some more.

Alright, I've updated with what I think might be the least disgusting option. I tried starting from a clean slate, and only modifying `TestIsolatedStoreImpl` to return some trivial `ParentHistogram` implementation that fits our needs for testing, but it got ugly pretty quickly. Definitely possible, just ugly (I'd need to have my own symbol table and stats cache, duplicating the ones I'm wrapping). So I came back to this instead, and just modified it to not store a real histogram.

From the perspective of not burdening production code with test-only concerns, it's not too bad, but it's not perfect either - I think it's 48 bytes of memory overhead per histogram, if I'm counting correctly, and a few extra instructions. It's just the best option I could come up with without a more significant refactor.

Let me know if this is ok.

mergeconflict commented:
/retest

mergeconflict commented:
/retest
